src/platform/i386/entry.S

```diff
@@ -110,14 +94,10 @@ restore_from_thd:
 	pushl %esp;		\
 	call fn;		\
 	addl $4, %esp;		\
-	RESTORE_REGS_GENERAL;	\
-	popl %ds; 		\
-	popl %es;		\
-	popl %fs;		\
-	popl %gs;		\
-	addl $4, %esp; 		\
-	sti;			\
-	iret
+	testl %eax, %eax;	\
+	jne restore_from_thd;	\
+	RESTORE_REGS_ALL;	\
```

Comment:
I was missing something.  You're actually doing exactly what I meant to be asking.  Again, great job.